### PR TITLE
Add query comparison tests for $eq

### PR DESCRIPTION
**Pull request by vic-tsang**
This PR contains:
- query comparison tests for $eq 

Ref:
- Issue #362

### CODE REVIEW

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_bson_wiring.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    QueryTestCase(
+        id="int64",
+        filter={"a": {"$eq": Int64(9999999999)}},
+        doc=[{"_id": 1, "a": Int64(9999999999)}, {"_id": 2, "a": Int64(1)}],
+        expected=[{"_id": 1, "a": Int64(9999999999)}],
+        msg="$eq with Int64 (long)",
+    ),
+    QueryTestCase(
+        id="decimal128",
+        filter={"a": {"$eq": Decimal128("1.5")}},
```
Comment (danielfrankcom) [inline documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_bson_wiring.py]: Could be `DECIMAL128_ONE_AND_HALF`

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_string.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+ALL_TESTS = STRING_TESTS
+
+
+@pytest.mark.parametrize("test", pytest_params(ALL_TESTS))
+def test_eq_string(collection, test):
+    """Parametrized test for $eq string-equality subtleties."""
+    collection.insert_many(test.doc)
+    result = execute_command(collection, {"find": collection.name, "filter": test.filter})
+    assertSuccess(result, test.expected, ignore_doc_order=True)
```
Comment (danielfrankcom) [inline documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_string.py]: I think the various calls (not just in this file) are not passing through the `msg` field from the cases which means it won't show up in errors.

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_operand_handling.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    collection.insert_many([{"_id": 1, "a": 1}])
+    result = execute_command(collection, {"find": collection.name, "filter": {"$eq": 5}})
+    assertFailureCode(
+        result,
+        BAD_VALUE_ERROR,
+        msg="$eq at query root is an unknown top-level operator",
+    )
+
+
+# (id, docs, value, expected) — the explicit {a: {$eq: value}} and implicit
```
Comment (danielfrankcom) [inline documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_operand_handling.py]: Style nit: it's quite difficult to follow the case definitions here given the lack of types and clarity on what the parameters of the array mean. Consider using something like this which is closer to the style used elsewhere and more explicit imo:

```python
# Property [Form Equivalence]: the implicit {a: v} form selects the same
# documents as the explicit {a: {$eq: v}} form for any non-regex operand.
# (Regex is the one operand where they diverge; see test_eq_regex.)
EQ_FORM_EQUIVALENCE_TESTS: list[QueryTestCase] = [
    QueryTestCase(
        id="scalar_int",
        filter={"a": {"$eq": 1}},
        doc=[{"_id": 1, "a": 1}, {"_id": 2, "a": 2}],
        msg="$eq int operand should select the same documents in both forms",
    ),
    QueryTestCase(
        id="string",
        filter={"a": {"$eq": "x"}},
        doc=[{"_id": 1, "a": "x"}, {"_id": 2, "a": "y"}],
        msg="$eq string operand should select the same documents in both forms",
    ),
    QueryTestCase(
        id="null_and_missing",
        filter={"a": {"$eq": None}},
        doc=[{"_id": 1, "a": None}, {"_id": 2}, {"_id": 3, "a": 1}],
        msg="$eq null operand should select the same documents in both forms",
    ),
    QueryTestCase(
        id="array",
        filter={"a": {"$eq": [1, 2]}},
        doc=[{"_id": 1, "a": [1, 2]}, {"_id": 2, "a": [1, 2, 3]}],
        msg="$eq array operand should select the same documents in both forms",
    ),
    QueryTestCase(
        id="embedded_document",
        filter={"a": {"$eq": {"b": 1}}},
        doc=[{"_id": 1, "a": {"b": 1}}, {"_id": 2, "a": {"b": 2}}],
        msg="$eq document operand should select the same documents in both forms",
    ),
]


@pytest.mark.parametrize("test", pytest_params(EQ_FORM_EQUIVALENCE_TESTS))
def test_eq_implicit_form_equivalence(collection, test):
    """Test explicit {a: {$eq: v}} and implicit {a: v} forms select the same documents."""
    collection.insert_many(test.doc)
    value = test.filter["a"]["$eq"]
    explicit = execute_command(collection, {"find": collection.name, "filter": test.filter})
    implicit = execute_command(collection, {"find": collection.name, "filter": {"a": value}})
    assertSuccess(
        explicit,
        implicit["cursor"]["firstBatch"],
        ignore_doc_order=True,
        msg=test.msg,
    )
```

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_edge_cases.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    result = execute_command(collection, {"find": collection.name, "filter": test.filter})
+    assertSuccess(result, test.expected, ignore_doc_order=True)
+
+
+@pytest.mark.parametrize("test", pytest_params(NAN_TESTS))
+def test_eq_edge_cases_nan(collection, test):
+    """Parametrized test for $eq edge cases involving NaN."""
+    collection.insert_many(test.doc)
+    result = execute_command(collection, {"find": collection.name, "filter": test.filter})
+    assertSuccessNaN(result, test.expected, ignore_doc_order=True)
```
Comment (danielfrankcom) [inline documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_edge_cases.py]: Can we use `assertResult` here with `pytest.approx` like we do elsewhere? I don't think this case needs its own dedicated method if the one above is swapped to `assertResult`. This seems to be using the older style still

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_string.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            {"_id": 2, "a": "a  b"},
+            {"_id": 3, "a": "ab"},
+        ],
+        expected=[{"_id": 1, "a": "a b"}],
+        msg="$eq treats whitespace as significant: 'a b' does not match 'a  b' or 'ab'",
+    ),
+]
+
+
+ALL_TESTS = STRING_TESTS
```
Comment (danielfrankcom) [inline documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_string.py]: Is this needed? Seems redundant

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_bson_wiring.py (hunk 1 above)
Comment (vic-tsang) [reply documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_bson_wiring.py]: replaced with DECIMAL128_ONE_AND_HALF

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_string.py (hunk 5 above)
Comment (vic-tsang) [reply documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_string.py]: removed

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_string.py (hunk 2 above)
Comment (vic-tsang) [reply documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_string.py]: fixed

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_operand_handling.py (hunk 3 above)
Comment (vic-tsang) [reply documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_operand_handling.py]: added

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_edge_cases.py (hunk 4 above)
Comment (vic-tsang) [reply documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_edge_cases.py]: converted to assertResult

File: documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_operand_handling.py (hunk 3 above)
Comment (danielfrankcom) [reply documentdb_tests/compatibility/tests/core/operator/query/comparison/eq/test_eq_operand_handling.py]: Was this added in a local commit? I'm not seeing it in the one which was pushed most recently

**Review by danielfrankcom — APPROVED** [review]

**Review by eerxuan — APPROVED** [review]
